README.md
WheelCMS Properties
=============

[![Build Status](https://travis-ci.org/wheelcms/wheelcms_properties.png?branch=master)](https://travis-ci.org/wheelcms/wheelcms_properties)
[![Coverage Status](https://coveralls.io/repos/wheelcms/wheelcms_properties/badge.png?branch=master)](https://coveralls.io/r/wheelcms/wheelcms_properties?branch=master)

Extensible properties for WheelCMS types


CHANGES.txt
v0.1, 2014-11-06 -- Initial public release


quicktest.py
from twotest.quicktest import QuickDjangoTest

if __name__ == '__main__':
    QuickDjangoTest(
        apps=("wheelcms_properties",),
        installed_apps=(
            'django.contrib.sites',
            'django.contrib.messages',
            'django.contrib.staticfiles',
            'django.contrib.humanize',

            'userena',
            'guardian',

            'south',
            'haystack',

            'wheelcms_axle',
            'wheelcms_properties',

            'drole',
            'granules',
            'taggit',
            'two.ol',
            'two.bootstrap',
            'twotest',
            'wheelcms_axle.tests',

        ),
        ROOT_URLCONF="wheelcms_axle.quicktest_urls",
        ANONYMOUS_USER_ID=-1,
        HAYSTACK_CONNECTIONS={'default':{'engine':'haystack.backends.simple_backend.SimpleEngine'}},
        AUTH_PROFILE_MODULE="wheelcms_axle.WheelProfile",
        CLEANUP_MEDIA=True,
        TEST_MEDIA_ROOT="/tmp/wheelcms_axle_test",
        USE_TZ=True,
        STATIC_URL='/',
        STATIC_ROOT='',
        CONTENT_LANGUAGES=(('en', 'English'), ('nl', 'Nederlands')),
        FALLBACK='en',
        LANGUAGE_CODE='en',
    )


wheelcms_properties/models.py
from operator import itemgetter

from django.db import models
from two.ol.base import json

from json import loads as load_json, dumps as dump_json

from wheelcms_axle.models import Configuration as BaseConfiguration
from wheelcms_axle.configuration import BaseConfigurationHandler

from wheelcms_axle.registries.configuration import configuration_registry

from wheelcms_axle.content import type_registry, Content
from wheelcms_axle.actions import action

class Properties(models.Model):
    main = models.ForeignKey(BaseConfiguration, related_name="spoke_properties")

class PropertyForm(models.Model):
    conf = models.ForeignKey(Properties, related_name="forms")
    name = models.CharField(max_length=60, blank=False)
    form = models.TextField(default="[]", blank=False)

    types = models.TextField(default="")

class PropertyFormData(models.Model):
    """ stored form data, per form / content instance """
    form = models.ForeignKey(PropertyForm, related_name="data")
    content = models.ForeignKey(Content, related_name="properties")

    properties = models.TextField(default="{}") # 'json'

"""
    List all known spokes
    Provide option to define additional properties on each of them

"""

class PropertiesConfigurationHandler(BaseConfigurationHandler):
    id = "spoke_properties"
    label = "Content Properties"
    model = Properties
    form = None

    def view(self, handler, instance):
        ## it's a SPA!
        return handler.template("wheelcms_properties/configure_properties.html")

    @action
    @json
    def formlist(self, handler, instance):
        ## sort?
        spokes=[dict(key=k, value=v.title) for (k, v) in type_registry.items()]

        spokes.sort(key=itemgetter('value'))
        return dict(spokes=spokes,
                    forms=[dict(name=f.name, id=f.id, spokes=f.types.split(","))
                           for f in PropertyForm.objects.filter(conf=instance)])

    @action
    @json
    def formdata(self, handler, instance):
        request = handler.request

        id = int(request.REQUEST.get('id', -1))
        if request.method == "POST":
            formdata = load_json(request.POST.get('data', '{}'))
            extra = load_json(request.POST.get('extra', '{}'))
            formname = extra.get('name', 'new form')
            spokes = ",".join(extra.get('spokes', []))

            if id == -1:
                pf = PropertyForm(conf=instance, name=formname,
                                  form=dump_json(formdata),
                                  types=spokes)
                pf.save()
                id = pf.id
            else:
                pf = PropertyForm.objects.get(pk=id)
                pf.name = formname
                pf.form = dump_json(formdata)
                pf.types = spokes
                pf.save()

        if id != -1:
            pf = PropertyForm.objects.get(pk=id)
            return dict(form=load_json(pf.form),
                        extra=dict(name=pf.name, id=pf.id,
                                   spokes=pf.types.split(",")))
        return dict(form=[], extra={'name':'', 'spokes':[]})

from wheelcms_axle.spoke import tab

from wheelcms_axle import permissions as p
from wheelcms_axle.actions import action_registry

def has_form(spoke):
    """ TODO: check if any forms have been assigned XXX """
    return True

@action(p.edit_content)
@json
def properties_data_handler(handler, request, action):
    """ return relevant form data """
    ## return first form that matches. TODO: combine forms into groups

    ## combine with data

    # import pdb; pdb.set_trace()
    spoke = handler.spoke()
    if request.method == "POST":
        data = load_json(request.POST.get('data'))
        for (id, formdata) in data.iteritems():
            form_obj_storage, _ = PropertyFormData.objects.get_or_create(form_id=id, content=spoke.instance)
            form_obj_storage.properties = dump_json(formdata)
            form_obj_storage.save()

    spokename = handler.spoke().name()
    forms = []
    for pf in PropertyForm.objects.all():
        types = pf.types.split(",")
        if spokename in types:
            forms.append(dict(id=pf.id, name=pf.name, form=load_json(pf.form)))

    data = {}
    for pfd in spoke.instance.properties.all():
        data[pfd.form_id] = load_json(pfd.properties)
    return dict(forms=forms, data=data)


@tab(p.edit_content, label="Object properties", condition=has_form)
def properties_handler(handler, request, action):
    """ only show tab if actions registered on spoke ? """
    return handler.template("wheelcms_properties/object_properties_tab.html")


def register():
    action_registry.register(properties_handler, 'properties')
    action_registry.register(properties_data_handler, 'properties_data')

    configuration_registry.register(PropertiesConfigurationHandler)

register()


wheelcms_properties/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from south.utils import datetime_utils as datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'Properties'
        db.create_table(u'wheelcms_properties_properties', (
            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('main', self.gf('django.db.models.fields.related.ForeignKey')(related_name='properties', to=orm['wheelcms_axle.Configuration'])),
        ))
        db.send_create_signal(u'wheelcms_properties', ['Properties'])

        # Adding model 'PropertyForm'
        db.create_table(u'wheelcms_properties_propertyform', (
            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('conf', self.gf('django.db.models.fields.related.ForeignKey')(related_name='forms', to=orm['wheelcms_properties.Properties'])),
            ('name', self.gf('django.db.models.fields.CharField')(max_length=60)),
            ('form', self.gf('django.db.models.fields.TextField')(default='[]')),
            ('types', self.gf('django.db.models.fields.TextField')(default='')),
        ))
        db.send_create_signal(u'wheelcms_properties', ['PropertyForm'])


    def backwards(self, orm):
        # Deleting model 'Properties'
        db.delete_table(u'wheelcms_properties_properties')

        # Deleting model 'PropertyForm'
        db.delete_table(u'wheelcms_properties_propertyform')


    models = {
        u'wheelcms_axle.configuration': {
            'Meta': {'object_name': 'Configuration'},
            'analytics': ('django.db.models.fields.CharField', [], {'default': "''", 'max_length': '50', 'blank': 'True'}),
            'description': ('django.db.models.fields.TextField', [], {'default': "''", 'blank': 'True'}),
            'head': ('django.db.models.fields.TextField', [], {'default': "''", 'blank': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'mailto': ('django.db.models.fields.EmailField', [], {'default': "''", 'max_length': '100', 'blank': 'True'}),
            'sender': ('django.db.models.fields.CharField', [], {'default': "''", 'max_length': '100', 'blank': 'True'}),
            'sendermail': ('django.db.models.fields.EmailField', [], {'default': "''", 'max_length': '100', 'blank': 'True'}),
            'theme': ('django.db.models.fields.CharField', [], {'default': "'default'", 'max_length': '256', 'blank': 'True'}),
            'title': ('django.db.models.fields.CharField', [], {'default': "''", 'max_length': '256', 'blank': 'True'})
        },
        u'wheelcms_properties.properties': {
            'Meta': {'object_name': 'Properties'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'main': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'properties'", 'to': u"orm['wheelcms_axle.Configuration']"})
        },
        u'wheelcms_properties.propertyform': {
            'Meta': {'object_name': 'PropertyForm'},
            'conf': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'forms'", 'to': u"orm['wheelcms_properties.Properties']"}),
            'form': ('django.db.models.fields.TextField', [], {'default': "'[]'"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '60'}),
            'types': ('django.db.models.fields.TextField', [], {'default': "''"})
        }
    }

    complete_apps = ['wheelcms_properties']

wheelcms_properties/static/js/wheelcms_properties.js
wheelcms_properties = angular.module('wheelcms_properties', ["schemaForm", "ui.sortable"]);

wheelcms_properties.factory('FormTool', function($sce) {
    var trust = $sce.trustAsHtml;
     // trust = function(s) { return s; };

    return {
        to_schema: function(e) {
            var s = {
                type: "object",
                properties: {}
            };
            var ee = {type: e.type, title: e.title};

            if(typeof e.choices != "undefined") {
                ee['enum'] = e.choices;
            }

            if(e.type == "number") {
                ee['default'] = parseInt(e.default_value, 10);
            }
            else if(e.type == "checkboxes") {
                ee['default'] = e.default_value.split("\n");
            }
            else {
                ee['default'] = e.default_value;
            }
            s.properties[e.id] = ee;

            return s;
        },

        to_form: function(e) {
            var field = {};

            field.key = e.id; // trust(e.id);
            field.type = (e.input || "text");

            if(e.choices) {
                field.titleMap = [];
                for(var c = 0; c < e.choices.length; c++) {
                    var cc = trust(e.choices[c]);
                    field.titleMap.push({value:e.choices[c], name:cc});
                }
            }

            return [ field ];
        },
        build_schemaform: function(formdef) {
            var schema = { type: "object", properties: {} };

            var form = [];

            for(var idx = 0; idx < formdef.length; idx++) {
                var e = formdef[idx];
                var s = this.to_schema(e);
                var f = this.to_form(e);

                schema.properties[e.id] = s.properties[e.id];
                form.push(f[0]);
            }
            return {schema: schema, form: form};
        }
    }
});

wheelcms_properties.controller("BaseFieldsEditCtrl", function($scope, $http, $modal, $filter, FormTool, Notifier) {
    $scope.changed = false;
    $scope.field_def = [];
    $scope.base = "";

    $scope.initialize = function(base) {
        $scope.base = base;
        $scope.model = {};

        return $http.get(base).then(function(result) {
            $scope.field_def = result.data.form;
            $scope.extra = result.data.extra;
            build_schemaforms();
        });
    };

    $scope.sortableOptions = {
        update: function(e, ui) {
            $scope.changed = true;
        }
    };

    var build_schemaforms = function() {
        $scope.schemaforms = [];
        for(var idx = 0; idx < $scope.field_def.length; idx++) {
            var e = $scope.field_def[idx];
            $scope.schemaforms.push({schema: FormTool.to_schema(e), form: FormTool.to_form(e),
                                     field: e, model: {}});
        }
    }

    build_schemaforms();


    $scope.newEditField = function(field) {
        var m = $modal.open({
                        templateUrl: 'AddEditField.html',
                        controller: 'AddEditFieldCtrl',
                        resolve: {
                            field: function() {
                                if(field !== null) {
                                    return field.field;
                                }
                            }
                        }
                });
        m.result.then(function(result) {
            if(field === null) {
                // add
                $scope.field_def.push(result);
                $scope.schemaforms.push({schema: FormTool.to_schema(result),
                                         form: FormTool.to_form(result),
                                         field: result,
                                         model: {}});
            }
            else {
                // update
                field.field = result;
                field.schema = FormTool.to_schema(result);
                field.form = FormTool.to_form(result);
                field.model = {};
            }

            $scope.changed = true;
        });
    };

    $scope.save = function() {
        var reordered = [];
        for(var idx = 0; idx < $scope.schemaforms.length; idx++) {
            reordered.push($scope.schemaforms[idx].field);
        }
        $http.post($scope.base,
                   {
                        data: $filter('json')(reordered),
                        extra: $filter('json')($scope.extra || {})
                   }).then(function(result) {
                             $scope.field_def = result.data.form;
                             $scope.extra = result.data.extra;
                             build_schemaforms();
                             $scope.changed = false;
                             Notifier.notify("info", "Saved");
        });
    }
});


wheelcms_properties.controller("WheelPropertiesCtrl",
                               function($scope, $controller, $rootScope,
                                        $http, FormTool) {
    $scope.state = "list";
    $scope.forms = [];

    var baseurl = $rootScope.urlbase + '?config=spoke_properties';

    var refreshforms = function() {
        $http.get(baseurl + "&action=formlist").then(function(result) {
            $scope.forms = result.data.forms;
            $scope.spokes = result.data.spokes;
        });
    }

    refreshforms();

    $controller('BaseFieldsEditCtrl', {$scope:$scope});

    /*
     * We need to save additional data on top of the base form:
     * - name for the form
     * - content types it's bound to
     * Either hook into the BaseFieldsEditCtrl edit/delete, or do a second get/post
     */
    var findspoke = function(key) {
        for(var j = 0; j < $scope.spokes.length; j++) {
            if($scope.spokes[j].key == key) {
                return $scope.spokes[j];
            }
        }
        return '';
    }

    $scope.findspoke = findspoke;

    $scope.newEditForm = function(id) {
        $scope.state = "edit";
        $scope.extra = {};

        var url = baseurl + '&action=formdata';
        if(id !== null) {
            url += "&id=" + id;
        }
        $scope.initialize(url).then(function(result) {
            $scope.extra.select_types = [];
            for(var i = 0; i < $scope.extra.spokes.length; i++) {
                var s = findspoke($scope.extra.spokes[i]);
                if(s) {
                    $scope.extra.select_types.push(s);
                }
            }
        });
    };
    $scope.cancelForm = function(id) {
        $scope.state = "list";
    };

    $scope.saveForm = function(id) {
        $scope.extra.spokes = [];
        for(var i = 0; i < $scope.extra.select_types.length; i++) {
            $scope.extra.spokes.push($scope.extra.select_types[i].key);
        }
        $scope.save(); // defined in base controller
        $scope.state = "list";
        refreshforms();
    };
});

wheelcms_properties.controller('AddEditFieldCtrl', function($scope, $modalInstance, field) {
    $scope.fielddata = angular.copy(field || {});

    $scope.data_types = [
            {title:"String", value:'string'},
            {title:"Number", value:'number'},
            {title:"Multiple", value:'checkboxes'}
        ];

    $scope.input_types = [
            {title:"Input", value:"text"},
            {title:"Password", value:"password"},
            {title:"Text area", value: "textarea"},
            {title:"Number", value: "number"},
            {title:"Checkbox", value: "checkbox"},
            {title:"Checkboxes", value: "checkboxes"},
            {title:"Select", value: "select"}
    ];

    for(var i = 0; i < $scope.data_types.length; i++) {
        if($scope.data_types[i].value == $scope.fielddata.type) {
            $scope.fielddata._type = $scope.data_types[i];
        }
    }
    for(var i = 0; i < $scope.input_types.length; i++) {
        if($scope.input_types[i].value == $scope.fielddata.input) {
            $scope.fielddata._input = $scope.input_types[i];
        }
    }

    if($scope.fielddata.choices) {
        $scope.fielddata._choices = $scope.fielddata.choices.join("\n")
    }

    $scope.ok = function () {
        // convert field/data selection back
        $scope.fielddata.type = $scope.fielddata._type.value || 'string';
        $scope.fielddata.input = $scope.fielddata._input.value || 'text';
        if($scope.fielddata._choices) {
            var choices = $scope.fielddata._choices.split("\n");
            $scope.fielddata.choices = [];
            for(var c = 0; c < choices.length; c++) {
                $scope.fielddata.choices.push(choices[c]);
            }
            if($scope.fielddata.choices.length == 0) {
                delete $scope.fielddata.choices;
            }
        }
        else {
            delete $scope.fielddata.choices;
        }
        $modalInstance.close($scope.fielddata);
    };

    $scope.cancel = function () {
        $modalInstance.dismiss('cancel');
    };
});

wheelcms_properties.controller("BaseFieldsViewCtrl", function($scope, $modal, $sce, $http, $rootScope, $filter, FormTool, Notifier) {
    var _url = null;

    $scope.forms = [];
    $scope.models = {};

    var load_devprops = function(data) {
        $scope.forms = [];


        $scope.models = data.data || {};
        for(var i = 0; i < data.forms.length; i++) {
            var form = data.forms[i];
            var sf = FormTool.build_schemaform(form.form);

            $scope.forms.push({id: form.id, schema:sf.schema, form:sf.form, model:{},
                               name:form.name});
        }
        console.log($scope.forms);
        console.log($scope.models);
    };

    $scope.initialize = function(url) {
        _url = url;
        $http.get(url).then(function(result) {
            console.log(result.data);
            load_devprops(result.data);
        });
    }


    $scope.save = function() {
        console.log($scope.models);
        console.log(_url);
        $http.post(_url,
                   {
                        data: $filter('json')($scope.models)
                   }).then(function(result) {
                        load_devprops(result.data);
                        Notifier.notify("info", "Saved");
                   });
    }
});

wheelcms_properties.controller("WheelPropertiesFieldsViewCtrl", function($scope, $rootScope, $controller) {
    $controller('BaseFieldsViewCtrl', {$scope:$scope});

    $scope.initialize($rootScope.urlbase + '+properties_data/');
});


wheelcms_properties/migrations/0002_auto__add_propertyformdata.py
# -*- coding: utf-8 -*-
from south.utils import datetime_utils as datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'PropertyFormData'
        db.create_table(u'wheelcms_properties_propertyformdata', (
            (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('form', self.gf('django.db.models.fields.related.ForeignKey')(related_name='data', to=orm['wheelcms_properties.PropertyForm'])),
            ('content', self.gf('django.db.models.fields.related.ForeignKey')(related_name='properties', to=orm['wheelcms_axle.Content'])),
            ('properties', self.gf('django.db.models.fields.TextField')(default='{}')),
        ))
        db.send_create_signal(u'wheelcms_properties', ['PropertyFormData'])


    def backwards(self, orm):
        # Deleting model 'PropertyFormData'
        db.delete_table(u'wheelcms_properties_propertyformdata')


    models = {
        u'auth.group': {
            'Meta': {'object_name': 'Group'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': u"orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        u'auth.permission': {
            'Meta': {'ordering': "(u'content_type__app_label', u'content_type__model', u'codename')", 'unique_together': "((u'content_type', u'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['contenttypes.ContentType']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        u'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Group']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "u'user_set'", 'blank': 'True', 'to': u"orm['auth.Permission']"}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        u'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        },
        u'wheelcms_axle.configuration': {
            'Meta': {'object_name': 'Configuration'},
            'analytics': ('django.db.models.fields.CharField', [], {'default': "''", 'max_length': '50', 'blank': 'True'}),
            'description': ('django.db.models.fields.TextField', [], {'default': "''", 'blank': 'True'}),
            'head': ('django.db.models.fields.TextField', [], {'default': "''", 'blank': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'mailto': ('django.db.models.fields.EmailField', [], {'default': "''", 'max_length': '100', 'blank': 'True'}),
            'sender': ('django.db.models.fields.CharField', [], {'default': "''", 'max_length': '100', 'blank': 'True'}),
            'sendermail': ('django.db.models.fields.EmailField', [], {'default': "''", 'max_length': '100', 'blank': 'True'}),
            'theme': ('django.db.models.fields.CharField', [], {'default': "'default'", 'max_length': '256', 'blank': 'True'}),
            'title': ('django.db.models.fields.CharField', [], {'default': "''", 'max_length': '256', 'blank': 'True'})
        },
        u'wheelcms_axle.content': {
            'Meta': {'object_name': 'Content'},
            'allowed': ('django.db.models.fields.TextField', [], {'default': 'None', 'null': 'True', 'blank': 'True'}),
            'classes': ('django.db.models.fields.related.ManyToManyField', [], {'symmetrical': 'False', 'related_name': "'content'", 'blank': 'True', 'to': u"orm['wheelcms_axle.ContentClass']"}),
            'created': ('django.db.models.fields.DateTimeField', [], {'null': 'True', 'blank': 'True'}),
            'description': ('django.db.models.fields.TextField', [], {'default': "''", 'blank': 'True'}),
            'discussable': ('django.db.models.fields.NullBooleanField', [], {'default': 'None', 'null': 'True', 'blank': 'True'}),
            'expire': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime(2034, 11, 14, 0, 0)', 'null': 'True', 'blank': 'True'}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'language': ('django.db.models.fields.CharField', [], {'max_length': '10'}),
            'meta_type': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'modified': ('django.db.models.fields.DateTimeField', [], {'null': 'True', 'blank': 'True'}),
            'navigation': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'node': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'contentbase'", 'null': 'True', 'to': u"orm['wheelcms_axle.Node']"}),
            'owner': ('django.db.models.fields.related.ForeignKey', [], {'to': u"orm['auth.User']", 'null': 'True'}),
            'publication': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now', 'null': 'True', 'blank': 'True'}),
            'state': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'template': ('django.db.models.fields.CharField', [], {'default': "''", 'max_length': '255', 'blank': 'True'}),
            'title': ('django.db.models.fields.CharField', [], {'max_length': '256'})
        },
        u'wheelcms_axle.contentclass': {
            'Meta': {'object_name': 'ContentClass'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '256'})
        },
        u'wheelcms_axle.node': {
            'Meta': {'object_name': 'Node'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'position': ('django.db.models.fields.IntegerField', [], {'default': '0'}),
            'tree_path': ('django.db.models.fields.CharField', [], {'default': "'0x3c88284e11e3bc33dL'", 'unique': 'True', 'max_length': '255'})
        },
        u'wheelcms_properties.properties': {
            'Meta': {'object_name': 'Properties'},
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'main': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'spoke_properties'", 'to': u"orm['wheelcms_axle.Configuration']"})
        },
        u'wheelcms_properties.propertyform': {
            'Meta': {'object_name': 'PropertyForm'},
            'conf': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'forms'", 'to': u"orm['wheelcms_properties.Properties']"}),
            'form': ('django.db.models.fields.TextField', [], {'default': "'[]'"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '60'}),
            'types': ('django.db.models.fields.TextField', [], {'default': "''"})
        },
        u'wheelcms_properties.propertyformdata': {
            'Meta': {'object_name': 'PropertyFormData'},
            'content': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'properties'", 'to': u"orm['wheelcms_axle.Content']"}),
            'form': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'data'", 'to': u"orm['wheelcms_properties.PropertyForm']"}),
            u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'properties': ('django.db.models.fields.TextField', [], {'default': "'{}'"})
        }
    }

    complete_apps = ['wheelcms_properties']